### ISSUE TITLE
malloc_usable_size/tc_malloc_size hangs under debug malloc when custom operator new implementation calls malloc_usable_size

### ISSUE THREAD

**Issue by gaurabpaul** [gperftools/gperftools#718]
attempting to use malloc_usable_size() with tcmalloc_debug simply hangs here:

#0  pthread_once () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_once.S:94
#1  0x00007f10b280bb27 in MallocExtension::instance () at src/malloc_extension.cc:207
#2  0x00007f10b28159e9 in tc_malloc_size (ptr=0x1010020) at src/debugallocation.cc:1419

(gdb) f 1
#1  0x00007f10b280bb27 in MallocExtension::instance () at src/malloc_extension.cc:207
207	  perftools_pthread_once(&module_init, InitModule);
(gdb) p module_init
$1 = 1
(gdb) p InitModule
$2 = {void (void)} 0x7f10b280b580 <InitModule()>
(gdb) f 0
#0  pthread_once () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_once.S:94
94		jmp	6b
(gdb) l
89		orl	%fs:PRIVATE_FUTEX, %esi
90	# endif
91	#endif
92		movl	$SYS_futex, %eax
93		syscall
94		jmp	6b
95
96		/* Preserve the pointer to the control variable.  */
97	3:	pushq	%rdi
98		cfi_adjust_cfa_offset(8)

any ideas on what could be going wrong here..

**Comment by alk** [comment]
Most likely some other thread is doing that MallocExtension::instance work and is stuck somewhere.

Please post backtraces of all threads. E.g. by telling gdb: thread apply all bt

**Comment by gaurabpaul** [comment]
the reason seems to be this: there are over-rides for "malloc" and "new" but NOT for "malloc_usable_size": both the malloc and new overrides use malloc_usable_size() . 

so in tcmalloc:  size_t malloc_usable_size(void* p) __THROW      ALIAS(tc_malloc_size);

extern "C" PERFTOOLS_DLL_DECL size_t tc_malloc_size(void* ptr) __THROW {
  return MallocExtension::instance()->GetAllocatedSize(ptr);
}

static void InitModule() {                                                                                                                                       current_instance = new MallocExtension;                                                                                                                       #ifndef NO_HEAP_CHECK                                                                                                                                           HeapLeakChecker::IgnoreObject(current_instance);                                                                                                              #endif                                                                                                                                                          │
}

so the malloc() override invokes malloc_usable_size() which invokes tc_malloc_size() which invokes 
"new" override which invokes malloc_usable_size() 

hence it is going into a recursive loop as above. 

the only way seems to be to override malloc_usable_size() so that 
current_instance = new MallocExtension; does not get invoked inside of tcmalloc_size()

implementing malloc_usable_size using the malloc_chunk_ptr from libc malloc fails with tcmalloc debug poisoning. 


any suggestions.

**Comment by alk** [comment]
Indeed. Thanks for reporting it.

Using placement new in this place looks like right fix. I.e. similar to: https://github.com/gperftools/gperftools/commit/185bf3fcc36f8cb3839abdfe652f615bfb5306d1

**Comment by alk** [comment]
Actually, I'm not able to see how tc_new may call tc_malloc_size under debug or production tcmalloc. Can you please post _full_ backtraces?

**Comment by alk** [comment]
Also please consider using latest version of tcmalloc since we're not using pthread_once in this place since gperftools 2.2

**Comment by gaurabpaul** [comment]
OK, we have done the over-ride for malloc and new which calls into malloc_usable_size. 

so our malloc() override invokes malloc_usable_size() which invokes tc_malloc_size() which in turn invokes our "new" override which invokes our malloc override and then it again invokes malloc_usable_size()

the following code in tcmalloc -> src/malloc_extension.cc 

204       current_instance = new MallocExtension;

is causing the recursive loop. 

gperftools 2.2 did not help either: it is breaking with the similar recursive loop here:

#0  tcmalloc::ThreadCache::InitModule () at src/thread_cache.cc:274
#1  0x00007ffff62c7d59 in tcmalloc::ThreadCache::GetCache () at src/thread_cache.h:408
#2  0x00007ffff62c0661 in (anonymous namespace)::do_malloc (size=size@entry=88) at src/tcmalloc.cc:1188
#3  0x00007ffff62e50fe in do_malloc (size=88) at src/sampler.h:146
#4  Allocate (type=-271733872, size=40) at src/debugallocation.cc:529
#5  DebugAllocate (type=-271733872, size=40) at src/debugallocation.cc:1027
#6  do_debug_malloc_or_debug_cpp_alloc (size=40) at src/debugallocation.cc:1200
#7  tc_malloc (size=40) at src/debugallocation.cc:1214
#8  0x0000000000429210 in alloc (alloc_func=0x4068f0 <__libc_malloc@plt>, size=8, instr_p=0x7ffff62d9e9e <InitModule()+14>) at ../../src/dfs-server/base/memtrack.cpp:134
#9  0x00000000004295d4 in operator new (size=8) at ../../src/dfs-server/base/memtrack.cpp:281
#10 0x00007ffff62d9e9e in InitModule () at src/malloc_extension.cc:204
#11 0x00007ffff62d9f99 in InitModule () at src/malloc_extension.cc:212
#12 MallocExtension::instance () at src/malloc_extension.cc:213
#13 0x00007ffff62e45e9 in tc_malloc_size (ptr=0x13385c0) at src/debugallocation.cc:1451

so how do we override tc_malloc_size() so that it does not call "new" ? since that is getting into this this recursive stack.

**Comment by alk** [comment]
Thanks. Now it makes sense. Placement new for MallocExtension should fix the problem.

Do you want to try doing that yourself ?

**Comment by gaurabpaul** [comment]
can please tell me the diff ?

**Comment by alk** [comment]
What diff you want me to tell ?

**Comment by gaurabpaul** [comment]
using the Placement new for MallocExtension  and where ?

**Comment by gaurabpaul** [comment]
actually using Placement new in 

tcmalloc -> src/malloc_extension.cc

204 current_instance = new MallocExtension;

will perhaps not work. since "new" is overridden to invoke malloc_usable_size () right ?

**Comment by alk** [comment]
It will work. Placement new will not call your implementation but just call constructor passing it whatever address is passed to placement new

**Comment by gaurabpaul** [comment]
ok, can you let me know the change in malloc_extension.cc:204 .thanks

**Comment by alk** [comment]
So thinking about this more...

We can certainly plug this specific issue quite easily. But a lot of things can really happen during initialization time. If not in current version, it can appear later.

I think your app would be safer it if starts with your logic disabled and then enables it during init time. Something along this lines:

```
bool setup_malloc_tracking() {
   // ensure things are inited
  char *rv = new char;
  malloc_usable_size(rv);
  delete rv;
  return true;
}

// false by default
static bool enable_malloc_tracking = setup_malloc_tracking();

void *operator new(size_t size) {
  void *rv = tc_new(size);
  //only account-for/check if we're sure malloc is fully initialized
  if (enable_malloc_tracking) {
     size_t real_size = malloc_usable_size(rv);
     if (!allow_alloc(real_size)) {
        delete rv;
        return maybe_throw_oom();
     }
  }
  return rv;
}
```

**Comment by gaurabpaul** [comment]
actually since we need to track every allocation and we have over-ridden the weak symbols the above solution does not seem to work as is; the same recursion appears as follows

#0  operator new (size=8)  [over-ride new]
#1  0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#2  0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#3  MallocExtension::instance () at src/malloc_extension.cc:213
#4  0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe6e060) at src/debugallocation.cc:1451
#5  0x000000000042907a in setup_malloc_tracking ()
#6  0x000000000042963a in operator new (size=8)  [over-ride new]
#7  0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#8  0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#9  MallocExtension::instance () at src/malloc_extension.cc:213
#10 0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe6e020) at src/debugallocation.cc:1451
#11 0x000000000042907a in setup_malloc_tracking()
#12 0x000000000042963a in operator new (size=8)  [over-ride new]
#13 0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#14 0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#15 MallocExtension::instance () at src/malloc_extension.cc:213
#16 0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe46020) at src/debugallocation.cc:1451
......
#19 0x00000000004292ce in malloc (size=568) at ../../src/dfs-server/base/memtrack.cpp:160
#20 0x00007ffff50bc4fd in __fopen_internal (filename=0x7fffffffd520 "/proc/net/psched", mode=0x7ffff4c355b0 "r", is32=1) at iofopen.c:73
#21 0x00007ffff4c0ecfe in get_psched_settings () from /lib64/libnl.so.1
#22 0x00007ffff7deb503 in call_init (env=0x7fffffffe5b8, argv=0x7fffffffe5a8, argc=1, l=<optimized out>) at dl-init.c:82
#23 _dl_init (main_map=0x7ffff7ffe208, argc=1, argv=0x7fffffffe5a8, env=0x7fffffffe5b8) at dl-init.c:131
#24 0x00007ffff7ddd45a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2

as long as we have "malloc" invoking "new" through the "malloc_usable_size()" we have this issue 

for malloc_extension.cc:204 InitModule() : 
 current_instance = new MallocExtension;

what are the other options __libc_malloc() (constructor init??)

**Comment by alk** [comment]
My point was that your code should _not_ call malloc_usable_size until initialization is complete. In this case, clearly this is not the case.

**Comment by gaurabpaul** [comment]
so you would want malloc_usable_size(rv); to be invoked at least once so that init goes through the first time right ? 

if that is true, it seems to be not possible with the following construct since malloc_usable_size() ends up invoking new () before it gets a chance to set the enabled_memtrack flag since we need to track every allocation.

from your example..

static bool enabled_memtrack = false;

void setup_memtrack() {
  // ensure things are inited
  char *rv = (char *)tc_new(sizeof(char));
  malloc_usable_size(rv);
  enabled_memtrack = true; 
  delete rv;
}

and malloc_usable_size() invokes malloc_extension.cc:204 InitModule() : 
current_instance = new MallocExtension;

void* operator new (std::size_t size) throw(std::bad_alloc)
{
    if (unlikely(!enabled_memtrack)) {
        setup_memtrack();
    }

   override new-code
}

is there explicit InitModule() call which is exposed by tcmalloc() that can be used in this situation ?

**Comment by alk** [comment]
No we won't export InitModule since it is implementation detail.

Also your code is far from what I've suggested. You're invoking infinite new recursion yourself in this code.

**Comment by gaurabpaul** [comment]
ok, in the above code, the recursion is as follows: 

#0  operator new (size=8) at 
#1  0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#2  0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#3  MallocExtension::instance () at src/malloc_extension.cc:213
#4  0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe460a0) at src/debugallocation.cc:1451
#5  0x000000000042908b in setup_memtrack () 
#6  0x0000000000429693 in operator new (size=8)
#7  0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#8  0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#9  MallocExtension::instance () at src/malloc_extension.cc:213
#10 0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe46060) at src/debugallocation.cc:1451
#11 0x000000000042908b in setup_memtrack () 
#12 0x0000000000429693 in operator new (size=8) 
#13 0x00007ffff5ea4e9e in InitModule () at src/malloc_extension.cc:204
#14 0x00007ffff5ea4f99 in InitModule () at src/malloc_extension.cc:212
#15 MallocExtension::instance () at src/malloc_extension.cc:213
#16 0x00007ffff5eaf5e9 in tc_malloc_size (ptr=0xe46020) at src/debugallocation.cc:1451
#17 0x000000000042908b in setup_memtrack () 
#18 0x00000000004292b7 in _libc_malloc@plt>, size=568, instr_p=0x7ffff50bc4fd <__fopen_internal+29>)
   #19 0x0000000000429327 in malloc (size=568) 
#20 0x00007ffff50bc4fd in __fopen_internal (filename=0x7fffffffd520 "/proc/net/psched", mode=0x7ffff4c355b0 "r", is32=1) at iofopen.c:73
#21 0x00007ffff4c0ecfe in get_psched_settings () from /lib64/libnl.so.1
#22 0x00007ffff7deb503 in call_init (env=0x7fffffffe5b8, argv=0x7fffffffe5a8, argc=1, l=<optimized out>) at dl-init.c:82
#23 _dl_init (main_map=0x7ffff7ffe208, argc=1, argv=0x7fffffffe5a8, env=0x7fffffffe5b8) at dl-init.c:131
#24 0x00007ffff7ddd45a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2

from the notes above, enabled_memtrack = true;  needs to be after malloc_usable_size(rv); has been invoked at least once right ? 

the recursive call to new() seems to be from malloc_usable_size() as seen above ..

**Comment by gaurabpaul** [comment]
and finally the following version of setup with tc_new or new overriddern by tc_new ends up with malloc_usable_size returning 0 

bool setup_malloc_tracking() {
  if (enable_malloc_tracking) {
    return true;
  }
   // ensure things are inited
  char *rv = (char *)tc_new(sizeof(char));
  size_t sz = malloc_usable_size(rv);
  assert(sz >= sizeof(char));
  free(rv);
  return true;
}

as it hits : 

 size_t MallocExtension::GetAllocatedSize(const void* p) {                                                                                                     assert(GetOwnership(p) != kNotOwned);                                                                                                                     return 0;                                                                                                                                                │
   }

it looks like malloc_usable_size() with tcmalloc() needs to be fixed for te above situation..

**Comment by alk** [comment]
At this point we're increasingly risking larger misunderstanding. We'll need small program that demonstrates issues that you're seeing.

On other note, it looks like you're trying to do that tracking even very early during program initialization. If true (and I'll be able to see that in your small test program), please note, that we are not supporting it.

At Couchbase we've been doing memore usage tracking via malloc hooks and consequently it is something that activates after main is called.

**Comment by gaurabpaul** [comment]
yes, our program simply overrides malloc/new/free (weak symbols). 

#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <assert.h>
#include <malloc.h>
#include <cstddef>
#include <new>

extern "C" {
  void* __libc_malloc(size_t size);
}

void *alloc(void *(*alloc_func)(size_t), const size_t size, const void *instr_p)
{
    void *ptr = (*alloc_func)(size);
    if (!ptr) {
        return ptr;
    }

    size_t sz = malloc_usable_size(ptr);
    assert(sz >= size);

    return ptr;
}

//! \brief  standard malloc substitution
void *malloc(size_t size)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}

void* operator new (std::size_t size) throw(std::bad_alloc)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}

int main() {
    printf("tcmalloc issue");
}

g++ tc.c -ltcmalloc_debug

and run it :  you have the recursion with tcmalloc and the core..

the recursive issue arises because new invokes malloc which invokes malloc_usable_size and then it in turn invokes new inside tcmalloc. 

so at couchbase the malloc hooks (__malloc_hook = my_malloc_hook)  thing work with tcmalloc  and how does that work for "new"?

**Comment by alk** [comment]
Can you attach the program instead? Or upload it somewhere? I.e. so that I don't have to deal with includes.

Regarding malloc hooks, I'm talking about https://github.com/gperftools/gperftools/blob/master/src/gperftools/malloc_hook_c.h#L72

**Comment by gaurabpaul** [comment]
attached ..
![tc](https://cloud.githubusercontent.com/assets/14847298/10126116/ebbbc4bc-653f-11e5-8929-80eb5be31dfc.png)

**Comment by alk** [comment]
Sorry, but your attachment doesn't work.

**Comment by gaurabpaul** [comment]
R u not able to download it. The code is the same as pasted above. (Just putting the braces in the #include)

**Comment by alk** [comment]
Not able

**Comment by gaurabpaul** [comment]
#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <assert.h>
#include <malloc.h>
#include <cstddef>
#include <new>

extern "C" {
  void* __libc_malloc(size_t size);
}

void *alloc(void *(*alloc_func)(size_t), const size_t size, const void *instr_p)
{
    void *ptr = (*alloc_func)(size);
    if (!ptr) {
        return ptr;
    }

    size_t sz = malloc_usable_size(ptr);
    assert(sz >= size);

    return ptr;
}

//! \brief  standard malloc substitution
void *malloc(size_t size)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}

void* operator new (std::size_t size) throw(std::bad_alloc)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}

int main() {
    printf("tcmalloc issue");
}

**Comment by gaurabpaul** [comment]
the above code (pasted again since attachment does not seem to work) reproduces the issue with tcmalloc . g++ tc.c -ltcmalloc_debug

**Comment by gaurabpaul** [comment]
also do not see ways to establish MallocHook:: hooks for memalign and its children. so the suggested hook-path also does not seem to work for all cases. 

other than malloc_usable_size (which maps to tc_malloc_size) is there any other interface (equivalent of GetAllocatedSize())  that is exposed. do not see. 

given a ptr 'X'  managed by tcmalloc, is there a particular offset from that pointer 'X' where the usable-size is stored (similar to libc malloc) ? if so, we can calculate the size reading bytes from that offset. (until this issue is fixed). please let me know.

**Comment by alk** [comment]
Please find some way to get code _verbatim_ to my end. Please don't make me cut and edit this thing.

BTW markdown does seem to support clean way to past code.

**Comment by gaurabpaul** [comment]
```
#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <assert.h>
#include <malloc.h>
#include <cstddef>
#include <new>


extern "C" {
  void* __libc_malloc(size_t size);
}


void *alloc(void *(*alloc_func)(size_t), const size_t size, const void *instr_p)
{
    void *ptr = (*alloc_func)(size);
    if (!ptr) {
        return ptr;
    }


    size_t sz = malloc_usable_size(ptr);
    assert(sz >= size);


    return ptr;
}


//! \brief  standard malloc substitution
void *malloc(size_t size)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}


void* operator new (std::size_t size) throw(std::bad_alloc)
{
    return alloc(__libc_malloc, size, __builtin_return_address(0));
}


int main() {
    printf("tcmalloc issue");
```

**Comment by gaurabpaul** [comment]
ok the above is verbatim.

**Comment by alk** [comment]
As I've described above malloc_usable_size will not work early during program initialization. Not that it's impossible to support, but I don't see compelling use-case for that and it has cost.

As I've advised above, just don't rely on malloc_usable_size until program initialization is complete.

Regarding hooks. Memalign does invoke new hook.

Regarding offsets. No metadata for small allocations is stored separately, there is no simple offset to read usable size from.
